### PR TITLE
#72 Entertainment page

### LINKED ISSUE #72: Запланированный досуг: просмотр и добавление
Blocker for #75 #77 #80 

After completing should be provided refactoring in #81 #82 #83 #84 

В работе пользуемся js библиотекой для googleAPI: [смотреть документацию](https://developers.google.com/maps/documentation/javascript/places?hl=ru)
Описание как работать с [картой](https://developers.google.com/maps/documentation/javascript?hl=ru), [поиском местом](https://developers.google.com/places/web-service?hl=ru)

Для работы нужно зарегистрированное приложение на гугл платформе - можете создать свое и добавить в переменные окружения или воспользоваться созданным раннее. по вопросам переменных окружения к @valerachuk 

Самый большой челендж - показать то что пользователь ввел вручную на карте.

При переходе на вкладку "Досуг" в боковом меню личного кабинета пользователя попадаем на страницу просмотра запланированного досуга
Также страничка доступна по адрессу `my/entertainments` 

страница запланированного досуга [исходник](https://www.figma.com/file/1WPfvOuZS9x7pwFlQbT1ht/%D0%9F%D1%80%D0%BE%D0%B5%D0%BA%D1%82-%D0%A2%D1%83%D1%80%D0%B8%D0%B7%D0%BC?node-id=1365%3A3479)
image

В рамках задачи необходимо реализовать:
- [x] отображение информации по досугу в виде плиток
- [x] переход на планирование досуга по нажатию на кнопку `Добавить`
- [x] просмотр на карте места досуга

Просмотр на карте осуществляется путем выбора места развлечения одним нажатием на плитку данного развлечения. При этом выбранная карточка выделяется другим цветом, а карта справа отображает местоположение выбранного развлечения. Используется гугл апи.
Пример выделения выбранной карточки (синеватый бордер с тенюшкой)
image

Фильтр, сортировка - выносится в отдельную задачу #75 

После нажатие на кнопку добавить на странице просмотра списка запланированного досуга, отображается форма для добавления досуга по адрессу `/entertainments/new`
[исходник](https://www.figma.com/file/1WPfvOuZS9x7pwFlQbT1ht/%D0%9F%D1%80%D0%BE%D0%B5%D0%BA%D1%82-%D0%A2%D1%83%D1%80%D0%B8%D0%B7%D0%BC?node-id=3408%3A16339)
image

Поля для планирования досуга:
- название - обязательное - валидация на непустоту и максимальную длину не больше 200 символов 
- дата и время - дата обязательное, время опционально (?) - может быть как будущим так и прошедшим
- адресс - обязательное
- ссылка - необязательное - валидация что это ссылка
- заметки - необязательное - ограничений 2000 символов
- добавить файл - необязательное - можно добавлять несколько файлов

Поля `название` и `адресс` в на страничке запланированного досуга - добавить подгружение результатов поиска мест через гугл апи [автоподстановки ](https://developers.google.com/maps/documentation/javascript/places-autocomplete?hl=ru)- может помочь получить `place_id` для его отображения на карте и получения дополнительной информации. [Подробнее ](https://developers.google.com/places/place-id) про `place_id`. Про сохранение айдишки - [читаем ](https://developers.google.com/places/place-id#save-id) про необходимость обновления. если пользователь не выбрал ничего из результатов поиска по гуглу - ориенируемся для карты на адресс.

После сохранения запланированного места досуга - показать уведомление об успешном сохранении и перенаправить на список.

По двойному нажатию по плитке досуга можно попасть на страницу редактирования `/entertainments/:id`, страница такая же как для добавления нового, но все поля заполнены ранее сохраненными значениями или полученными автоматически

### PR DESCRIPTION
**Pull request by ivhorodko**

### CODE REVIEW

File: front/src/components/mmt-leisure-card/mmt-leisure-card.component.css [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+.close {
+  color: var(--leisure-place-close-state-color);
+}
+
+.decription {
+  margin-top: 6px;
+}
+
+.active {
+  border: 1px solid #2962FF;
```
Comment (horba) [inline front/src/components/mmt-leisure-card/mmt-leisure-card.component.css]: css variable

File: front/src/components/mmt-leisure-card/mmt-leisure-card.component.html [hunk 2]
```diff
@@ -0,0 +1,62 @@
+<div :class="`${ isActive ? 'active' : '' } place`">
```
Comment (horba) [inline front/src/components/mmt-leisure-card/mmt-leisure-card.component.html]: `<div class="place" :class="{ active: isActive }">`

https://ru.vuejs.org/v2/guide/class-and-style.html#%D0%98%D1%81%D0%BF%D0%BE%D0%BB%D1%8C%D0%B7%D0%BE%D0%B2%D0%B0%D0%BD%D0%B8%D0%B5-%D0%BE%D0%B1%D1%8A%D0%B5%D0%BA%D1%82%D0%BE%D0%B2

https://michaelnthiessen.com/dynamically-add-class-name/

Comment (horba) [reply front/src/components/mmt-leisure-card/mmt-leisure-card.component.html]: place -> place-card ?

File: front/src/components/mmt-leisure-card/mmt-leisure-card.component.html [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+  <div class="photo-wrap">
```
Comment (horba) [inline front/src/components/mmt-leisure-card/mmt-leisure-card.component.html]: wrap - is a verb not noun - photo-wrapper

File: front/src/components/mmt-leisure-card/mmt-leisure-card.component.html [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    <img
+      v-if="place.photo"
+      class="photo"
+      :src="place.photo"
+    >
+    <img
+      v-else
+      class="photo"
+      src="http://www.rangerwoodperiyar.com/images/joomlart/demo/default.jpg"
```
Comment (horba) [inline front/src/components/mmt-leisure-card/mmt-leisure-card.component.html]: add to assets source of default photo for leisure card as default-leisure-photo.jpg

File: front/src/components/mmt-leisure-card/mmt-leisure-card.component.html [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    >
+  </div>
+  <div class="place-info">
+    <router-link :to="`/my/entertainments/manage/${place.id}`" class="icon">
```
Comment (horba) [inline front/src/components/mmt-leisure-card/mmt-leisure-card.component.html]: manage -> schedule?

File: front/src/components/mmt-leisure-card/mmt-leisure-card.component.html [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    </router-link>
+    <div class="place-body-wrap">
+      <div class="title-wrap">
+        <h6
+          class="title"
+          :title="place.title"
+        >
+          {{ place.title |  trimString(20) }}
+        </h6>
+        <div class="rating-wrap" v-if="place.rating">
```
Comment (horba) [inline front/src/components/mmt-leisure-card/mmt-leisure-card.component.html]: wrapper

File: front/src/components/mmt-leisure-card/mmt-leisure-card.component.html [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+      class="photo"
+      src="http://www.rangerwoodperiyar.com/images/joomlart/demo/default.jpg"
+    >
+  </div>
+  <div class="place-info">
+    <router-link :to="`/my/entertainments/manage/${place.id}`" class="icon">
+      <v-icon small>mdi-lead-pencil</v-icon>
+    </router-link>
+    <div class="place-body-wrap">
+      <div class="title-wrap">
```
Comment (horba) [inline front/src/components/mmt-leisure-card/mmt-leisure-card.component.html]: wrapper

File: front/src/components/mmt-leisure-card/mmt-leisure-card.component.html [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+        >
+          {{ place.title |  trimString(20) }}
+        </h6>
+        <div class="rating-wrap" v-if="place.rating">
+          <p class="rating">{{ place.rating }}</p>
+        </div>
+      </div>
+    </div>
+    <div class="place-footer-wrap">
+      <div class="schedule-wrap">
```
Comment (horba) [inline front/src/components/mmt-leisure-card/mmt-leisure-card.component.html]: wrapper

File: front/src/components/mmt-leisure-card/mmt-leisure-card.component.js [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+  },
+  name: 'mmt-leisure-card',
+  props: ['place', 'isActive'],
+  filters: {
+    splitTypes (value) {
+      return value.join(', ');
+    },
+    trimString (value, count) {
+      return value.length > count ? value.substr(0, count) + '...' : value;
+    }
```
Comment (horba) [inline front/src/components/mmt-leisure-card/mmt-leisure-card.component.js]: make common general filters for joining and trinmming

File: front/src/store/modules/entertainment.store.js [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+      return api.post('/assets/entertainment', formData,
+        {
+          headers: {
+            'Content-Type': 'multipart/form-data'
+          }
+        });
+    },
+    deleteFile ({ commit }, fileName) {
+      return api.delete(`/assets/entertainment/${fileName}`);
+    }
```
Comment (horba) [inline front/src/store/modules/entertainment.store.js]: might, to add functionality for uploading files into api service?

File: front/src/views/user/entertainments-new/entertainments-new.view.css [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+
+.datepicker >>> .v-label {
+  top: -6px;
+  left: -22px !important;
+  color: black;
+}
+
+.datepicker >>> .v-icon {
+  cursor: pointer;
+} */
```
Comment (horba) [inline front/src/views/user/entertainments-new/entertainments-new.view.css]: no drafts

File: front/src/views/user/entertainments-new/entertainments-new.view.html [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+        transition="scale-transition"
+        offset-y
+        max-width="290px"
+        min-width="290px"
+      >
+        <template v-slot:activator="{ on, attrs }">
+          <v-text-field
+            prepend-inner-icon="mdi-calendar-check"
+            v-model="visitDate"
+            label="Дата"
```
Comment (horba) [inline front/src/views/user/entertainments-new/entertainments-new.view.html]: translations

File: front/src/views/user/entertainments-new/entertainments-new.view.js [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+        lat: 48.459322,
+        lng: 35.052729
+      },
+      markerPos: null,
+      placeTitle: '',
+      peopleCount: '',
+      note: '',
+      place: null,
+      menu: false,
+      visitDate: new Date().toISOString().substr(0, 10)
```
Comment (horba) [inline front/src/views/user/entertainments-new/entertainments-new.view.js]: we have filter for date formatting based on date-fns - it has sense to use it

File: front/src/views/user/entertainments-new/entertainments-new.view.js [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+  },
+  components: {
+    VDatePicker,
+    VMenu,
+    VTextField,
+    VBtn,
+    VIcon,
+    MmtLeisureCard,
+    VTextarea,
+    MmtTextInput
```
Comment (horba) [inline front/src/views/user/entertainments-new/entertainments-new.view.js]: order
VTextArea,
MmtLeisureCard,
MmtLeisureCard

File: front/src/views/user/entertainments-new/entertainments-new.view.css [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+
+.mmt-btn.v-btn:not(.v-btn--disabled):not(.v-btn--flat):not(.v-btn--text):not(.v-btn--outlined) >>> .v-btn__content {
+  color: var(--continue-button-active-color);
+}
+
+.mmt-btn {
+  padding:10px 48px;
+  margin: 14px 0;
+  color: white;
+}
```
Comment (horba) [inline front/src/views/user/entertainments-new/entertainments-new.view.css]: might, create one more type of button and styles for it in common styles?
like text-button or other one?

File: front/src/views/user/entertainments-new/entertainments-new.view.css [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+  margin: 0 20px 0 40px;
+}
+
+.new-entertainments-map {
+  width: 100%;
+  height: 100%;
+}
+
+.new-entertainments-autocomplete {
+  width: 94%;
```
Comment (horba) [inline front/src/views/user/entertainments-new/entertainments-new.view.css]: why peecents?

File: front/src/views/user/entertainments/entertainments.view.html [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+      </span>
+    </div>
+    <div class="buttons-container">
+      <v-btn
+        rounded
+        depressed
+        class="details accent lighten-1 mr-5"
+        to="/my/entertainments/manage">
+        <span class="text-capitalize">
+          Добавить
```
Comment (horba) [inline front/src/views/user/entertainments/entertainments.view.html]: translations

File: front/src/views/user/entertainments/entertainments.view.html [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+        </span>
+      </v-btn>
+      <v-btn
+        rounded
+        depressed
+        class="details accent lighten-1"
+      >
+        <span class="text-capitalize">
+          Фильтр
```
Comment (horba) [inline front/src/views/user/entertainments/entertainments.view.html]: translations

File: server/Entities/EntertainmentRepository.cs [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+namespace Entities
+{
+  public class EntertainmentRepository
+  {
+    private readonly RepositoryContext RepositoryContext;
+
+    public EntertainmentRepository(RepositoryContext repositoryContext)
+    {
+      RepositoryContext = repositoryContext;
+    }
```
Comment (horba) [inline server/Entities/EntertainmentRepository.cs:21]: don't we you interface for repository?

File: server/Entities/EntertainmentRepository.cs [hunk 20]
```diff
@@ -0,0 +1,39 @@
+﻿using Entities.Models;
+using System.Linq;
+
+namespace Entities
+{
+  public class EntertainmentRepository
+  {
+    private readonly RepositoryContext RepositoryContext;
```
Comment (horba) [inline server/Entities/EntertainmentRepository.cs]: IRepositoryContext _repositoryContext

File: server/Entities/Models/Entertainment.cs [hunk 21]
```diff
@@ -0,0 +1,27 @@
...
+    [ForeignKey("User")]
+    public int UserId { get; set; }
+    public string PlaceId { get; set; }
+    public string Title { get; set; }
+    public string Note { get; set; }
+    public int PeopleCount { get; set; }
+    public DateTime VisitDate { get; set; }
+    public double? LocationLat { get; set; }
+    public double? LocationLng { get; set; }
+    public string EntertainmentFilePath { get; set; }
```
Comment (horba) [inline server/Entities/Models/Entertainment.cs:24]: required properties and other validation rules for database?

File: server/WebAPI/Controllers/EntertainmentController.cs [hunk 22]
```diff
@@ -0,0 +1,57 @@
...
+
+namespace WebAPI.Controllers
+{
+  [Route("api/[controller]")]
+  [ApiController]
+  [Authorize]
+  public class EntertainmentController : ControllerBase
+  {
+
+    private readonly EntertainmentService _entertainmentService;
```
Comment (horba) [inline server/WebAPI/Controllers/EntertainmentController.cs]: IEntertainmentService _entertainmentService

use DI

File: server/WebAPI/Services/EntertainmentService.cs [hunk 23]
```diff
@@ -0,0 +1,47 @@
...
+using System.Linq;
+using WebAPI.DTO;
+
+namespace WebAPI.Services
+{
+  public class EntertainmentService
+  {
+
+    private readonly IMapper _mapper;
+    private readonly EntertainmentRepository _entertainmentRepository;
```
Comment (horba) [inline server/WebAPI/Services/EntertainmentService.cs]: IEntertainmentRepository _entertainmentRepository

File: server/WebAPI/Services/Assets/AssetType.cs [hunk 24]
```diff
@@ -2,6 +2,7 @@
 {
   public enum AssetType
   {
-    UserAvatar = 1
+    UserAvatar = 1,
+    Entertainment = 3
```
Comment (horba) [inline server/WebAPI/Services/Assets/AssetType.cs:7]: where 2?

Comment (ivhorodko) [reply server/WebAPI/Services/Assets/AssetType.cs:7]: Because Irina uses 2 in her pull request

Comment (horba) [reply server/WebAPI/Services/Assets/AssetType.cs:7]: and what type is it?

**Review by horba — DISMISSED** [review]

**Review by horba — APPROVED** [review]

**Pull request merged**